### PR TITLE
perf(core): replace unconditional txn sync with a conditional one. 10x improvement of ingress perf for small transactions

### PR DESCRIPTION
**Pull request by bluestreak01**
- remove hard "msync" from critical path (e.g. commit). "msync" is conditional on database being in SYNC mode
- remove read-lock when WAL writer is activated. When metadata version doesn't change WAL activation follows fast path
- refactor out caching of `cairo.commit.mode` value, so it can be reloaded without server restart

PG Wire 2.0 bench before and after. 10 second run of inserts, two fields, commit every row (batch size = 1).

Before:

```
    2,196 inserts    14,042 selects
    4,496 inserts    43,795 selects
    6,760 inserts    75,253 selects
    9,022 inserts   107,121 selects
   11,044 inserts   137,401 selects
   13,246 inserts   170,169 selects
   15,388 inserts   201,575 selects
   17,636 inserts   234,439 selects
   19,868 inserts   266,073 selects
   22,090 inserts   295,647 selects

12 JDBC inserters, 1 selectors combined performance:
2,207 inserts per second, 29,545 selects per second
Per-thread performance:
183 inserts per second, 29,545 selects per second
```

After:
```
  166,154 inserts    19,153 selects
  363,517 inserts    49,326 selects
  535,017 inserts    78,953 selects
  689,515 inserts   109,986 selects
  827,700 inserts   140,584 selects
  957,144 inserts   172,022 selects
1,073,416 inserts   204,529 selects
1,184,451 inserts   237,443 selects
1,287,076 inserts   271,185 selects
1,380,068 inserts   301,670 selects

12 JDBC inserters, 1 selectors combined performance:
137,958 inserts per second, 30,156 selects per second
Per-thread performance:
11,496 inserts per second, 30,156 selects per second
```

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 50 / 50 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 66 / 66 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/Vm.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CsvFileIndexer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryPMARImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 68 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/Vm.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CsvFileIndexer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryPMARImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 68 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/Vm.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CsvFileIndexer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryPMARImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 90 (96.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|14|16|87.50%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/Vm.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CsvFileIndexer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryPMARImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|14|14|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 89 / 92 (96.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerAPI.java|14|16|87.50%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV2.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/IndexBuilder.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalEventWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/Vm.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableSequencerImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/SequencerMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/SimpleReadWriteLock.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CsvFileIndexer.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryPMARImpl.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLog.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/seq/TableTransactionLogV1.java|14|14|100.00%|

File: core/src/main/java/io/questdb/cairo/vm/MemoryPMARImpl.java [hunk 1]
```diff
@@ -138,8 +139,8 @@ public void switchTo(FilesFacade ff, long fd, long extendSegmentSize, long offse
     }
 
     public void sync(boolean async) {
-        if (pageAddress != 0 && commitMode != CommitMode.NOSYNC) {
-            ff.msync(pageAddress, getPageSize(), commitMode == CommitMode.ASYNC);
+        if (pageAddress != 0) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/cairo/vm/MemoryPMARImpl.java:142]: is this change intentional? this will cause `msync()` even when `CairoConfig` says NOSYNC

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/vm/MemoryPMARImpl.java:142]: yeah, it is intentional. The logic is when signature is `sync(boolean)` we are committing, the question is not "if" it is "how". This function is called from a site, that has a check if NOSYNC is set or not.
